In [4]:
import pandas as pd

# Loading Data

In [12]:
url = 'https://raw.github.com/pandas-dev/pandas/main/pandas/tests/io/data/csv/iris.csv'
df = pd.read_csv(url)
df.head(15)

SepalLength  SepalWidth  PetalLength  PetalWidth         Name
0           5.1         3.5          1.4         0.2  Iris-setosa
1           4.9         3.0          1.4         0.2  Iris-setosa
2           4.7         3.2          1.3         0.2  Iris-setosa
3           4.6         3.1          1.5         0.2  Iris-setosa
4           5.0         3.6          1.4         0.2  Iris-setosa
5           5.4         3.9          1.7         0.4  Iris-setosa
6           4.6         3.4          1.4         0.3  Iris-setosa
7           5.0         3.4          1.5         0.2  Iris-setosa
8           4.4         2.9          1.4         0.2  Iris-setosa
9           4.9         3.1          1.5         0.1  Iris-setosa
10          5.4         3.7          1.5         0.2  Iris-setosa
11          4.8         3.4          1.6         0.2  Iris-setosa
12          4.8         3.0          1.4         0.1  Iris-setosa
13          4.3         3.0          1.1         0.1  Iris-setosa
14          5.8         4.0          1.2         0.2  Iris-setosa

In [13]:
df.describe()

SepalLength  SepalWidth  PetalLength  PetalWidth
count   150.000000  150.000000   150.000000  150.000000
mean      5.843333    3.054000     3.758667    1.198667
std       0.828066    0.433594     1.764420    0.763161
min       4.300000    2.000000     1.000000    0.100000
25%       5.100000    2.800000     1.600000    0.300000
50%       5.800000    3.000000     4.350000    1.300000
75%       6.400000    3.300000     5.100000    1.800000
max       7.900000    4.400000     6.900000    2.500000

The features of this dataset will be the measurements of the flowers which are the first 3 columns. Once the flower's measurement is plotted this data will show which flower these types of measurements belong to. The type of flower will be the label. The mapping of the labels will classify which flower family the measurement belongs to.

# Splitting Data

In [23]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.utils import shuffle
df = shuffle(df)

X = df.iloc[:,:-1]
y = df.iloc[:,-1]
X_train, X_test, y_train, y_test = train_test_split(X,y,train_size=0.9,test_size=0.1,random_state=42)

# Logistic Regression

In [24]:
from sklearn.linear_model import LogisticRegression

log_reg = LogisticRegression()
log_reg.fit(X_train,y_train)

log_reg.score(X_test,y_test)

0.9333333333333333

This score for the logistic regression means that logistic regression is a really good way to classify which type of flower we are looking at. Logistic Regression seems to be a very good type of model to use whe are trying to use classify something for a dataset. 

In [47]:
print(log_reg.predict_proba(X_test.head(1)))
log_reg.classes_

[[1.17258812e-06 3.12626456e-02 9.68736182e-01]]


array(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype=object)

For the first dataset in the X_test dataset, you can see that this data point has the highest chance to be identified as a viriginica flower type. The chances of this datapoint being classified as the other 2 are way lower than the first class. The third class has a 96% while the second has around a 3% and the first one has a super low chance.

In [44]:
log_reg.coef_

array([[-0.48402288,  0.85004545, -2.38910086, -1.05089957],
       [ 0.56021955, -0.47252684, -0.15961602, -0.98763573],
       [-0.07619667, -0.37751861,  2.54871688,  2.0385353 ]])

In [45]:
log_reg.intercept_

array([ 10.07268986,   2.31651617, -12.38920603])

# SVM

In [56]:
from sklearn.svm import SVC

svm_model = SVC(kernel = 'linear', probability = True)
svm_model.fit(X_train, y_train)

svm_model.score(X_test,y_test)

1.0

In [57]:
svm_model.predict_proba(X_test.head(1))

array([[0.0027849 , 0.00486525, 0.99234984]])

After the model was fitted with the training dataset, it shows that the testing score comes out to be 100%. This means that when a this model guesses one of these three flowers it will always be correct. So far it seems like the SVM model does the best a multiple feature classifications, beating the Logisic Regression model by 6% using the same datasets.
The probability that the first datapoint in X_test will be guessed the third flower (virginica) is at 99%. This model is very confident at classfiying this dataset.

# Neural Network

In [131]:
from keras.models import Sequential
from keras.layers import Dense, Activation, Dropout
from keras.utils import to_categorical
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder

encode = LabelEncoder()
y_trainN = encode.fit_transform(y_train)
y_testN = encode.fit_transform(y_test)

nn_model = Sequential()
nn_model.add(Dense(20, activation = 'relu', input_shape = (4,)))
nn_model.add(Dense(10, activation = 'relu'))
nn_model.add(Dense(1, activation = 'softmax'))
nn_model.compile(loss = 'categorical_crossentropy', optimizer = 'adam', metrics = ['accuracy'])

nn_model.fit(X_train, y_trainN, epochs=20)

Epoch 1/20
5/5 [==============================] - 0s 1ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 2/20
5/5 [==============================] - 0s 1ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 3/20
5/5 [==============================] - 0s 3ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 4/20
5/5 [==============================] - 0s 2ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 5/20
5/5 [==============================] - 0s 1ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 6/20
5/5 [==============================] - 0s 2ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 7/20
5/5 [==============================] - 0s 3ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 8/20
5/5 [==============================] - 0s 2ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 9/20
5/5 [==============================] - 0s 3ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 10/20
5/5 [==============================] - 0s 4ms/step - loss: 0.0000e+00 - accuracy: 0.3333

In [124]:
test_acc = nn_model.evaluate(X_test,y_testN)

1/1 [==============================] - 0s 987us/step - loss: 0.0000e+00 - accuracy: 0.3333


In [132]:
nn_model = Sequential()
nn_model.add(Dense(100, activation = 'relu', input_shape = (4,)))
nn_model.add(Dense(10, activation = 'relu'))
nn_model.add(Dense(1, activation = 'softmax'))
nn_model.compile(loss = 'categorical_crossentropy', optimizer = 'adam', metrics = ['accuracy'])

nn_model.fit(X_train, y_trainN, epochs=50, batch_size = 4, verbose =1 )

Epoch 1/50
34/34 [==============================] - 0s 1ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 2/50
34/34 [==============================] - 0s 1ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 3/50
34/34 [==============================] - 0s 1ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 4/50
34/34 [==============================] - 0s 2ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 5/50
34/34 [==============================] - 0s 1ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 6/50
34/34 [==============================] - 0s 1ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 7/50
34/34 [==============================] - 0s 2ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 8/50
34/34 [==============================] - 0s 2ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 9/50
34/34 [==============================] - 0s 2ms/step - loss: 0.0000e+00 - accuracy: 0.3333
Epoch 10/50
34/34 [==============================] - 0s 1ms/step - loss: 0.0000e+0

In [133]:
test_acc = nn_model.evaluate(X_test,y_testN)

1/1 [==============================] - 0s 2ms/step - loss: 0.0000e+00 - accuracy: 0.3333


The Neural Networks seem to be not very good at finding the correct classficaion of the iris dataset. This neural networks accruacy score seems to only come out at around 33% which is a lot less than the other models. Having more hidden layers and more epochs did not seem to bring the model to predict better and stayed at 33% across all of the epochs which is kind of weird. The highest score achieved was 33%.

# K-Nearest Neighbor

In [60]:
from sklearn.neighbors import KNeighborsClassifier

knn_model = KNeighborsClassifier(n_neighbors=3)
knn_model.fit(X_test,y_test)



KNeighborsClassifier(n_neighbors=3)

In [61]:
knn_model.predict_proba(X_test.head(1))

array([[0., 0., 1.]])

In [84]:
knn_model.score(X_test,y_test)

1.0

This score means that when the K-Nearets Neighbor algorithm was looking for datapoints around this first datapoint in the X_test dataset. It was found that the 3 other datapoints close to the current one maps to the third class which is iris-virignica. The other 2 classes did not have any datapoints close to the first datapoint of the X_test dataset. The KNN porobability does not function the same as the other models because KNN is just trying to map the current datapoint to its closest datapoints to predict which class this datapoint belongs to.

# Conclusion

From all of the models above it seems that the Support Vector Machine was one of the most successful at accurately classfying a datapoint that was have no been trained yet. The SVM model prodicted all of the data in the X_test dataset correctly and matched exactly to the results shown in the y_test. The other models were also really close. The Logistic Regression model was getting a score of around 93% and this is already really great with a minor errors, but people would probably want a little possible error as possile in their mchine learning model, so this would not be the model to use. 
The Neural Network Models were not as good as the rest of models, so it would not be recommended to use this model, this does not mean that this model is bad and could probably be used way better at a different kind of dataset.
The KNN model was also super good at predicting the datapoints too, probably because it is using datapoints around itself to guess the current datapoint. The KNN model also had a 100% accuray as predicting the class of the flowers in the test data. This model is as useful as the SVM model and could be also used to help people classify the type of flowers in this datset without any errors.